xtrack/base_element.py

```diff
@@ -284,35 +331,37 @@ def _arr2ctx(self, arr):
 
 class PerParticlePyMethod:
 
-    def __init__(self, kernel, element):
-        self.kernel = kernel
+    def __init__(self, kernel_name, element):
+        self.kernel_name = kernel_name
         self.element = element
 
     def __call__(self, particles, increment_at_element=False, **kwargs):
+        instance = self.element
+        context = instance.context
+        if not hasattr(instance, '_track_kernel'):
+            if instance._track_kernel_name not in context.kernels.keys():
+                instance.compile_kernels(particles_class=particles.__class__)
+            instance._track_kernel = context.kernels[instance._track_kernel_name]
+        self.kernel = context.kernels[self.kernel_name]
```

Comment:
Please add test similar to this one
https://github.com/xsuite/xtrack/blob/a65498cca524f0c8092fb723a1eaf63d45ca9735/tests/test_elements.py#L875

ensuring that per-particle kernels work also for different Particles classes